src/Arcus.Messaging.ServiceBus.Core/Extensions/MessageExtensions.cs

```diff
@@ -1,20 +1,66 @@
-﻿using Arcus.Messaging.Abstractions;
+﻿using System;
+using System.Collections.Generic;
+using Arcus.Messaging.Abstractions;
 using GuardNet;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
 
 // ReSharper disable once CheckNamespace
 namespace Microsoft.Azure.ServiceBus
 {
     public static class MessageExtensions
     {
         /// <summary>
-        ///     Gets the transaction id that is linked to this message
+        ///     Gets the user property with a given <paramref name="key"/>
         /// </summary>
-        /// <param name="message">Message to process</param>
-        /// <returns>Transaction id for message</returns>
-        public static string GetTransactionId(this Message message)
+        /// <typeparam name="TProperty">The type of the value of the property.</typeparam>
+        /// <param name="message">The message to get the user property from.</param>
+        /// <param name="key">The key in the dictionary to look up the user property.</param>
+        public static TProperty GetUserProperty<TProperty>(this Message message, string key)
+        {
+            if (message.UserProperties.TryGetValue(key, out object value))
+            {
+                if (value is TProperty typed)
+                {
+                    return typed;
+                }
+
+                throw new InvalidCastException(
+                    $"The found user property with the key: '{key}' in the Service Bus message was not of the expected type: '{typeof(TProperty).Name}'");
+            }
+
+            throw new KeyNotFoundException(
+                $"No user property with the key: '{key}' was found in the Service Bus message");
+        }
+
+        /// <summary>
+        /// Append the <see cref="MessageCorrelationInfo"/> after the message is received.
```

Comment:
This is very vague, how about this:
```suggestion
        /// Gets the correlation information for a given message.
```